### PR TITLE
More snapshot abstraction

### PR DESCRIPTION
**Pull request by brson**
###  What have you changed?

This restructures some of the engine_traits traits to avoid recursive associated types, and then pulls up more of the snapshot abstractions through generic tikv functions. After this PR, considerably more of tikv is parameterized over engines.

Fixes https://github.com/tikv/tikv/issues/6124

Part of https://github.com/tikv/tikv/issues/4184

To avoid situations like:

```
trait KvEngine {
    type Snapshot: Snapshot;
}

trait Snapshot {
    type KvEngine: KvEngine
}
```

which result in generics that can't be typechecked because the two KvEngine instances can't be proved the same, "branch" traits are instead parameterized over their "parent" traits, like

```
trait KvEngine {
    type Snapshot: Snapshot;
}

trait Snapshot<E> where E: KvEngine { }
```

And then implementations only implement the traits for their concrete types like

```
impl Snapshot<RocksEngine> for RocksSnapshot { }
```

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

cargo test --all --no-run

###  Refer to a related PR or issue link (optional)

https://github.com/tikv/tikv/issues/6124

https://github.com/tikv/tikv/issues/4184

### CODE REVIEW

File: components/engine_traits/src/lib.rs [hunk 1]
```diff
@@ -147,6 +147,13 @@
 //!   - https://github.com/rust-lang/rfcs/pull/1598
 //!   - https://github.com/rust-lang/rust/issues/44265
 //!
+//! - Traits can't have mutually-recursive associated types. That is, if
+//!   `KvEngine` has a `Snapshot` associated type, `Snapshot` can't then have a
+//!   `KvEngine` associated type - the compiler will not be able to resolve both
+//!   `KvEngine`s to the name type. In these cases, e.g. `Snapshot` needs to be
```
Comment (nrc) [inline components/engine_traits/src/lib.rs]:
```suggestion
//!   `KvEngine`s to the same type. In these cases, e.g. `Snapshot` needs to be
```

**Comment by brson** [comment]
Updated. Comments addressed.

**Comment by brson** [comment]
Merge conflicts resolved.

**Pull request merged**